### DISCUSSION TITLE
Pattern Matching / Variable Binding in LINQ Query Syntax

### DISCUSSION THREAD

**Discussion by Mafii** [dotnet/csharplang#6799] [Language Ideas] [answered]
When using Ternary expressions or ifs (non-exhaustive list), a bound variable in a pattern match can be used in the "valid" branch:

```csharp
object value = ...;
var x = value is Type type ? type /* can be used here  */ : null /* not here */;
// or
if (value is Type type)
{
    // type can be used here
}
```

This is really nice, and can help to avoid unnecessary casts. Before this feature, casting `value` to `Type` would have to be done every time or a temporary variable had to be assigned to avoid multiple casts. It is in fact so useful that Java has added a similar feature by now and is planning to also provide it in switches (something C# already has!). So much to the status quo.

When using the LINQ query syntax, the bound variable is limited to the current where clause:

```csharp
from x in xs
where x is Type type
select type; // does not work right now
```

Workaround:

```csharp
from x in xs
where x is Type
let type = (Type)x // avoidable, useless cast
select type; // cast could be done here, but if any additional logic follows using type, a temporary variable is valuable for readability
```

When using the method syntax, `OfType<T>` can be used.

But that's besides the point - switching to the method syntax you lose the advantages of the query syntax - in some cases this is not an option as it simplifies things a lot. Using method syntax would make the code almost unreadable.

I am aware that the query syntax is simply syntactic sugar for select, select many, order by, etc.

## Proposal:

Right now let statement is desugared into a property (or is it a field?) on an anonymous type that is passed to the next select statement and can be used directly from following query parts.

A pattern match in a where should be desugared into the equivalent of `let <bound name> = (cast)variable`. In my opinion, this follows the principle of "least astonishment" as something similar is already the behavior for ternary expressions and ifs.

```csharp
from x in xs
where x is Type type
select type;
```

turns into (must not be an actual compiler step):

```csharp
from x in xs
where x is Type
let type = (Type)x
select type;
```

which is broadly the same as

```csharp
xs
  .Where(x => x is Type)
  .Select(x => new { x, type = (Type)x })
  .Select(anon => anon.type);
```

### Why not translate to `OfType<T>`?
More complex patterns are not translatable to `OfType<T>` directly, so adding that special case would add little extra value over solving this in a more general way.

**Comment by HaloFour** [comment, marked as answer]
See: #159 #355

I can't find the design meeting notes but I recall that the team rejected this and later ungated the use of patterns in LINQ expressions which would make it a breaking change to modify the scoping now.

**Reply by ufcpp** [reply]
https://github.com/dotnet/csharplang/discussions/4937
https://github.com/dotnet/csharplang/blob/main/meetings/2018/LDM-2018-01-03.md#scoping-of-expression-variables-in-query-clauses

**Comment by Mafii** [comment]
Unfortunate, I think the then-thought-of syntax wasn't extremely good, so I'm glad it wasn't introduced - having that feature in where's as proposed would be nice. What about gate-keeping that feature behind a new lang-version? And accepting the break just like with discard `_`?

**Reply by HaloFour** [reply]
> What about gate-keeping that feature behind a new lang-version? And accepting the break just like with discard `_`?

This is very unlikely to be considered as the team doesn't like introducing dialects of the language or to silently change the meaning of existing code.  Even with discards, the language will always continue to consider `_` to be a proper identifier in any case where it was legal code in prior versions.
